ondemand_agentmanager.go

```diff
@@ -72,60 +77,84 @@ func (m *OnDemandAgentManager) Reconfigure(opts OnDemandAgentManagerReconfigureO
 
 // GetClusterAgent returns the Agent which is not bound to any bucket.
 func (m *OnDemandAgentManager) GetClusterAgent() (*Agent, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
+	m.stateLock.Lock()
 	if m.closed {
+		defer m.stateLock.Unlock()
 		return nil, errors.New("agent manager closed")
 	}
+	defer m.stateLock.Unlock()
 
 	return m.clusterAgent, nil
 }
 
 // GetBucketAgent returns the Agent bound to a specific bucket, creating a new Agent as required.
 func (m *OnDemandAgentManager) GetBucketAgent(ctx context.Context, bucketName string) (*Agent, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
+	m.stateLock.Lock()
 	if m.closed {
+		defer m.stateLock.Unlock()
 		return nil, errors.New("agent manager closed")
 	}
+	defer m.stateLock.Unlock()
 
-	if m.bucketAgents == nil {
-		m.bucketAgents = make(map[string]*Agent)
-	} else {
-		bucketAgent, ok := m.bucketAgents[bucketName]
+	fastMap := m.fastMap.Load()
+	if fastMap != nil {
+		agent, ok := (*fastMap)[bucketName]
 		if ok {
-			return bucketAgent, nil
+			return agent, nil
 		}
 	}
 
-	bucketAgent, err := m.makeAgentLocked(ctx, bucketName)
+	agent, err := m.getBucketAgentSlow(ctx, bucketName)
+	if err != nil {
+		return nil, err
+	}
+
+	return agent, nil
+}
+
+func (m *OnDemandAgentManager) getBucketAgentSlow(ctx context.Context, bucketName string) (*Agent, error) {
+	m.slowLock.Lock()
+	defer m.slowLock.Unlock()
+	if agent, ok := m.slowMap[bucketName]; ok {
+		return agent, nil
+	}
+
+	agent, err := m.makeAgent(ctx, bucketName)
 	if err != nil {
 		return nil, err
 	}
+	m.slowMap[bucketName] = agent
 
-	m.bucketAgents[bucketName] = bucketAgent
+	fastMap := make(map[string]*Agent, len(m.slowMap))
+	for name, agent := range m.slowMap {
+		fastMap[name] = agent
```

Comment:
Why are we making a new fast map, with the contents of the slow map, instead of appending the newly created agent to the existing fast map?

Reply:
@Westwooo There are two reasons that our existing fastMaps do this (and thus I imagine Charlie followed that pattern).  One of the primary reasons is less of a technical one and more of a "defensive coding" sort of thing.  Because the fastMap critically _must_ be identical to the fastMap, we want to design the code such that it makes it unlikely that we end up with mistakes around the contents of the fastMap.  You can imagine if we add an item to the slowMap, but somehow fail to add the item to the fastMap, it means that the fastMap will be 'silently incorrect' forever.  By having the fastMap always be rebuilt it means it's extremely unlikely we could ever end up with a fastMap that's not equivalent to the slowMap.  Additionally, if we ever were to have the fastMap fall out of sync with the slowMap, the next time a modification was made to the structure, it would be 'repaired' through the rebuild process.  The second reason that this is set up this way is that because the fastMap is being atomically swapped out, it must be immutable once its placed into that atomic (or we would have race conditions around writes occuring to the fastMap while someone is trying to read it).  Since the fastMap needs to be immutable, and thus has to be a new object for each write, coupled with the first defensive coding point, it makes sense to simply build the structure from the 'source of truth' (the slowMap) each time rather than trying to copy the existing fastMap and try to make changes.

Reply:
Makes a lot of sense, thanks for the explanation